### DISCUSSION TITLE
How to define distinct schemas using single table approach

### DISCUSSION THREAD

**Discussion by CrystalMethod** [dynamoose/dynamoose#1140] [Q&A] [answered]
The following is taken from the online documentation:
https://dynamoosejs.com/guide/Model#dynamoosemodeldocumentname-schema
```
Behind the scenes Dynamoose will automatically pick the closest schema to match to your document,
and use that schema for all operations pertaining to that document. If no matching schema can be found,
it will default to the first schema in the array.
```
How does this mechanism work?

Given the following three schemas: ShapeSchema, ColorSchema, ShapeColorSchema:
```ts
export const ShapeSchema = new Schema(
  {
    pk: {
      type: String,
      hashKey: true,
      get: (value) => (value as string).slice('A#'.length),
      set: (value) => 'A#' + value,
    },
    sk: {
      type: String,
      rangeKey: true,
      index: {
        global: true,
        rangeKey: 'pk',
      },
      get: (value) => (value as string).slice('A#'.length),
      set: (value) => 'A#' + value,
    },
    shape: {
      type: String,
      required: true,
    },
  },
);
```

```ts
export const ColorSchema = new Schema(
  {
    pk: {
      type: String,
      hashKey: true,
      get: (value) => (value as string).slice('B#'.length),
      set: (value) => 'B#' + value,
    },
    sk: {
      type: String,
      rangeKey: true,
      index: {
        global: true,
        rangeKey: 'pk',
      },
      get: (value) => (value as string).slice('B#'.length),
      set: (value) => 'B#' + value,
    },
    color: {
      type: String,
      required: true,
    },
  },
);
```

```ts
export const ShapeColorSchema = new Schema(
  {
    pk: {
      type: String,
      hashKey: true,
      get: (value) => (value as string).slice('A#'.length),
      set: (value) => 'A#' + value,
    },
    sk: {
      type: String,
      rangeKey: true,
      index: {
        global: true,
        rangeKey: 'pk',
      },
      get: (value) => (value as string).slice('B#'.length),
      set: (value) => 'B#' + value,
    },
    shape: {
      type: String,
      required: true,
    },
    color: {
      type: String,
      required: true,
    },
  },
);
```

When storing `ShapeColor` items:

```json
{
  "pk": "123",
  "sk": "456",
  "shape": "circle",
  "color": "red" 
}
```

I typically get entries like the following:

```json
{
  "pk": "A#123",
  "sk": "A#456",
  "shape": "circle"
}
```
This leads me to the assumption that the wrong schema was used for serialisation.

**Comment by fishcharlie** [comment, marked as answer]
@CrystalMethod 

> How does this mechanism work?

High level.

It will basically look through each of your schema's, and assign a score to each:

1 - All type checks match
0.5 - Properties exist in document that don't exist in schema
0 - Type check fails

The first schema in your array you pass into your model with highest score will be used.

> This leads me to the assumption that the wrong scheme was used for serialisation

Yes looks like it. My guess is that it's part of #1133. That PR hasn't been included in a release _yet_. It has been merged tho and will be included in the next release. Hoping that'll happen really soon. But in the meantime you can point to the `master` branch to get that fix.

**Reply by CrystalMethod** [reply]
Thanks for you quick reply. I was able to test the master branch instead of v2.7.0. But the results are unfortunately the same. There is a small change if you sort the entries in the "schema" array differently.

All my document types have a "descriminator value attribute" - something very "old school". Could I somehow use this to help select the right schema? 😄

**Reply by fishcharlie** [reply]
@CrystalMethod 

> I was able to test the master branch instead of v2.7.0. But the results are unfortunately the same.

Ok this is a bug. Please submit a bug report for it. Please note the commit hash of the version you are running in the `Dynamoose Version` field.

> There is a small change if you sort the entries in the "schema" array differently.

That is expected. As I mentioned: `The first schema in your array you pass into your model with highest score will be used.`. However, the documentation might not be clear here. If you believe it's not, please submit another bug report for that.

> All my document types have a "descriminator value attribute" - something very "old school". Could I somehow use this to help select the right schema? 😄

Probably.

https://dynamoosejs.com/guide/Schema#attribute-types

There is a `Constant` type. Below is an example for it.

```js
{
    "id": String,
    "schemaType": {
        "type": {
            "value": "Constant",
            "settings": {
                "value": "Shape" // Any `schemaType` attribute must equal `Hello World` now.
            }
        }
    }
}
```

So you just use that `Constant` type and change the value for each of your schemas. That should cause type check to fail for some schemas and therefore raise priority of others.

**Reply by CrystalMethod** [reply]
Finally I got it working, even with 2.7.0. Your reference to the use of "Constant" schema attribute type was key. All schemas now contain a unique type attribute based on the discriminator value, as described above.
In addition, I have defined the schemas as precisely as possible: a)`enum` instead of just `string` if applicable and b) marked as "required" as much as possible. (not sure if this has some performance impacts).
Nevertheless, a wrong scheme was still used for "joined types". It was always the schema of one of the two simple types. (ShapeSchema + ColorSchema = ShapeColorSchema) So I finally had to adjust the order of the array of schemas so that "joined types" come first and then the simple types at the end. As you mentioned above: `The first schema in your array you pass into your model with highest score will be used.` And the schema of a simple type will always match a joined type (at least if all attributes of a simple type are part of the joined type - which is true in my case)

Thanks for you help!
